File: solutions/go/circular-buffer/1/circular_buffer.go
package circular

import "errors"

type Buffer struct {
	data     []byte
	size     int
	capacity int
	head     int
	tail     int
}

func NewBuffer(size int) *Buffer {
	return &Buffer{
		data:     make([]byte, size),
		capacity: size,
	}
}

func (b *Buffer) ReadByte() (byte, error) {
	if b.size == 0 {
		return 0, errors.New("cannot read from an empty buffer")
	}

	read := b.data[b.head]
	b.head = (b.head + 1) % b.capacity
	b.size--
	return read, nil
}

func (b *Buffer) WriteByte(c byte) error {
	if b.size == b.capacity {
		return errors.New("cannot write to a full buffer")
	}

	b.data[b.tail] = c
	b.tail = (b.tail + 1) % b.capacity
	b.size++
	return nil
}

func (b *Buffer) Overwrite(c byte) {
	if b.size == b.capacity {
		b.ReadByte()
	}
	b.WriteByte(c)
}

func (b *Buffer) Reset() {
	b.size = 0
	b.head = 0
	b.tail = 0
}

File: solutions/python/circular-buffer/1/circular_buffer.py
class BufferFullException(BufferError):
    """Exception raised when CircularBuffer is full.

    message: explanation of the error.

    """

    def __init__(self, message):
        self.message = message


class BufferEmptyException(BufferError):
    """Exception raised when CircularBuffer is empty.

    message: explanation of the error.

    """

    def __init__(self, message):
        self.message = message


class CircularBuffer:
    def __init__(self, capacity):
        self.__buffer = [0] * capacity
        self.capacity = capacity
        self.head, self.tail = 0, 0
        self.size = 0

    def read(self):
        if self.size == 0:
            raise BufferEmptyException("Circular buffer is empty")

        read_val = self.__buffer[self.head]
        self.head = (self.head + 1) % self.capacity
        self.size -= 1
        return read_val

    def write(self, data):
        if self.size == self.capacity:
            raise BufferFullException("Circular buffer is full")

        self.__buffer[self.tail] = data
        self.tail = (self.tail + 1) % self.capacity
        self.size += 1

    def overwrite(self, data):
        if self.size == self.capacity:
            self.read()
        self.write(data)

    def clear(self):
        self.size = 0
        self.head = self.tail
